packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/SyncSessionImpl.kt

```diff
@@ -15,10 +15,111 @@
+        return waitForChanges(TransferDirection.DOWNLOAD, timeout)
+    }
+
+    override suspend fun uploadAllLocalChanges(timeout: Duration): Boolean {
+        return waitForChanges(TransferDirection.UPLOAD, timeout)
+    }
+
+    /**
+     * Wrap Core callbacks that will not be invoked until data has been either fully uploaded
+     * or downloaded.
+     *
+     * When this method returns. The user facing Realm has been updated to the latest state.
+     *
+     * @param direction whether data is being uploaded or downloaded.
+     * @param timeout timeout parameter.
+     * @return `true` if the job completed before the timeout was hit, `false` otherwise.
+     */
+    private suspend fun waitForChanges(direction: TransferDirection, timeout: Duration): Boolean {
+        // Currently `realm` is only `null` when a SyncSession is created for use inside a
+        // ErrorHandler, and we expect this to be the only place, so it is safe to spell this
+        // out in the error message.
+        if (realm == null) {
+            throw IllegalStateException(
+                """
+                Uploading and downloading changes is not allowed when inside 
+                a `SyncSession.ErrorHandler`.
+                """.trimIndent()
+            )
+        }
+        require(timeout.isPositive()) {
+            "'timeout' must be > 0. It was: $timeout"
+        }
+
+        // Channel to work around not being able to use `suspendCoroutine` to wrap the callback, as
+        // that results in the `Continuation` being frozen, which breaks it.
+        val channel = Channel<Any>(1)
+        try {
+            val result: Any = withTimeout(timeout) {
+                withContext(realm.configuration.notificationDispatcher) {
+                    val callback = object : SyncSessionTransferCompletionCallback {
+                        override fun invoke(error: SyncErrorCode?) {
+                            if (error != null) {
+                                channel.trySend(SyncException(error.toString()))
+                            } else {
+                                channel.trySend(true)
+                            }
+                        }
+                    }.freeze()
+                    when (direction) {
+                        TransferDirection.UPLOAD -> {
+                            RealmInterop.realm_sync_session_wait_for_download_completion(nativePointer, callback)
+                        }
+                        TransferDirection.DOWNLOAD -> {
+                            RealmInterop.realm_sync_session_wait_for_upload_completion(nativePointer, callback)
+                        }
+                    }
+                    channel.receive()
+                }
+            }
+            if (direction == TransferDirection.DOWNLOAD) {
+                realm.refresh()
+            }
+            when (result) {
+                is Boolean -> return result
+                is Throwable -> throw result
```

Comment:
Should this throw a new exception with `cause = result` 🤔

Reply:
Hmm, I would think that propagating the underlying exception would be the correct thing to do? What exception would otherwise make sense here?

Reply:
It was mostly to got the full trace, but rethinking it might be the right thing just to have the original.